django/contrib/admin/options.py

```diff
@@ -1807,6 +1807,7 @@ def _changeform_view(self, request, object_id, form_url, extra_context):
                     return self.response_change(request, new_object)
             else:
                 form_validated = False
+                form.set_initial_data()
```

Comment:
This can't be right. 

Either we're solving a general problem in forms — in which case the solution needs to be applied inside `django/forms` — or else it's a specific problem in the admin — in which case the `set_initial_data` method would belong in `django/contrib/admin`, and not `django/forms`. 

As it is, we're neither one way nor the other. 

It **looks** like (from the test) you're suggesting it's a general problem with forms... in which case can we expand the tests to show the issue more clearly? 

Imagined test case would take a form, show its initial data, fail validation, and then show that initial data had changed. 🤔 With that in hand I think we could make a better decision, or a make that decision more easily.

Reply:
@carltongibson 
From what I undsertand, this situation now doens't belong to `either django/forms` or `django/contrib/admin` right ?
I think I can extend the tests. I'll reply back when I get my hands on the extended tests

Thank, you.

Reply:
@carltongibson 
Hi, cartongibson.

> Imagined test case would take a form, show its initial data, fail validation, and then show that initial data had changed. 🤔 With that in hand I think we could make a better decision, or a make that decision more easily.

I extended the tests like below.
```py
def test_form_data_when_fail_validation(self):
    class PersonForm(forms.Form):
        name = CharField(show_hidden_initial=True)
        extra = JSONField(show_hidden_initial=True)

    data = QueryDict(mutable=True)
    data.update(
        {
            "initial-name": "John",
            "name": "",
            "initial-extra": json.dumps({"height": 999, "weight": 99}),
            "extra": json.dumps({"height": 177, "weight": 70}),
        }
    )
    data._mutable = False

    form = PersonForm(data=data)

    previous_data = form.data
    # fail validation
    form.is_valid()
    after_data = form.data

    # Test succeed
    assert previous_data == after_data
    assert previous_data['initial-extra'] == after_data['initial-extra']
```
Through this test, it can be seen that the `initial-data` doesn't change even if the form fails validation.
However, it was found that this issue is a problem that occurs when the form is rendered like `as_table()` or `as_p()`.

Refer to the following code.
Model for test
```py
# tests.admin_custom_urls.models
# I added PersonInfo model for test.
...
class PersonInfo(models.Model):
    def json_empty(self):
        return ["none"]

    name = models.CharField(max_length=20)
    extra = models.JSONField(default=json_empty)


class PersonInfoAdmin(admin.ModelAdmin):
    def response_post_save_add(self, request, obj):
        return HttpResponseRedirect(
            reverse("admin:admin_custom_urls_personinfo_history", args=[obj.pk])
        )

site.register(PersonInfo, PersonInfoAdmin)
```

test code
```py
# tests.admin_custom_urls.tests
# I reproduced the issue of the ticket.
# And I removed `form.set_initial_data()` my code.

def test_post_save_change_fail_from_validation(self):
    PersonInfo.objects.create(name="Junki Yoon", extra=["test"])
    self.assertEqual(PersonInfo.objects.count(), 1)
    person_info = PersonInfo.objects.all()[0]
    post_url = reverse(
        "admin_custom_urls:admin_custom_urls_personinfo_change", args=[person_info.pk]
    )
    # fail validation.
    invalid_data = {"name": "", "extra":  json.dumps(["test123"]), "initial-extra": json.dumps(["test"])}
    response = self.client.post(post_url, invalid_data)

    print(response.context["adminform"].form.data)
    # output
    # <QueryDict: {'name': [''], 'extra': ['["test123"]'], 'initial-extra': ['["test"]']}>
    print(response.context["adminform"].form)
    # output
    # ...
    # <textarea name="extra" cols="40" rows="10" required id="id_extra">
    # [&quot;test123&quot;]</textarea>
    # <input type="hidden" name="initial-extra" value="[&quot;test123&quot;]" id="initial-id_extra">

   # Not equal. form.data["initial-extra"] != rendered_data's initial-extra value
```
Referring to the abvoe test, it can be seen that the `initial-extra` of `form.data` is not changed (as confirmed above test).
However, after the form in rendered the `initial-extra `is changed to the value of `extra`. I think this is the problem here.

From the perspective of the admin page.
1. Fail validation
2. The form is re-rendered while outputting an error message(This field is required.)
3. At this time `initial-data` is changed to the value of `extra`.

So I think it's a general problem in forms. (while rendering)

Reply:
@carltongibson 
Hello. cartongibson :) 

I clearly found it to be a problem with `bound_data` of `JsonField`. (I further debugged with what was mentioned in the previous comment.)

**Commit :** [49afce1](https://github.com/django/django/pull/16019/commits/49afce189bc15531ae567a95b7114f85c95e190e)
Added a paramater `only_initial` and If it is `True` it returns `initial` rather than `data`, So that the `initial-extra` field maintains the initial value even after form validation fails.

**Test Code**
In the test, We reproduced the situation where issue occured, and verfied that the `inital-extra` value did not change.